### PR TITLE
Support output descriptors in scantxoutset

### PR DESCRIPTION
**Pull request by sipa**
As promised, here is an implementation of my output descriptor concept (https://gist.github.com/sipa/e3d23d498c430bb601c5bca83523fa82) and integration within the `scantxoutset` RPC that was just added through #12196.

It changes the RPC to use descriptors for everything; I hope the interface is simple enough to encompass all use cases. It includes support for P2PK, P2PKH, P2WPKH, P2SH, P2WSH, multisig, xpubs, xprvs, and chains of keys - combined in every possible way.

### CODE REVIEW

File: src/script/descriptor.cpp [hunk 1]
```diff
@@ -0,0 +1,615 @@
+#include <script/descriptor.h>
```
[the remark this answers is not in the document]
Comment (sipa) [reply src/script/descriptor.cpp]: Fixed.

File: src/script/sign.h [hunk 2]
```diff
@@ -28,6 +28,20 @@ class SigningProvider
...
+struct FlatSigningProvider final : public SigningProvider
+{
+    std::map<CScriptID, CScript> scripts;
+    std::map<CKeyID, CPubKey> pubkeys;
+    std::map<CKeyID, CKey> keys;
+
+    bool GetCScript(const CScriptID& scriptid, CScript& script) const override;
+    bool GetPubKey(const CKeyID& keyid, CPubKey& pubkey) const override;
+    bool GetKey(const CKeyID& keyid, CKey& key) const override;
+
```
Comment (promag) [inline src/script/sign.h]: nit, remove empty line.

File: src/script/sign.h [hunk 3]
```diff
@@ -28,6 +28,20 @@ class SigningProvider
...
+};
+
+FlatSigningProvider Merge(const FlatSigningProvider& a, const FlatSigningProvider& b);
```
Comment (promag) [inline src/script/sign.h]: At the moment this is only used in tests, move there?

File: src/script/descriptor.cpp [hunk 4]
```diff
@@ -0,0 +1,619 @@
...
+    std::string ToString() const override { return OriginString() + ":" + m_provider->ToString(); }
+    bool ToPrivateString(const SigningProvider& arg, std::string& ret) const override
+    {
+        std::string sub;
+        if (!m_provider->ToPrivateString(arg, sub)) return false;
+        ret = OriginString() + ":" + std::move(sub);
+        return true;
+    }
+};
+#endif
```
Comment (promag) [inline src/script/descriptor.cpp]: nit, add `// KEY_ORIGIN_SUPPORT`.

File: src/script/descriptor.cpp [hunk 5]
```diff
@@ -0,0 +1,619 @@
...
+namespace {
+
+////////////////////////////////////////////////////////////////////////////
+// Internal representation                                                //
+////////////////////////////////////////////////////////////////////////////
+
+typedef std::vector<uint32_t> KeyPath;
+
+/** Interface for public key objects in descriptors. */
+struct PubkeyProvider {
```
Comment (promag) [inline src/script/descriptor.cpp]: nit, `{` in new line.

File: src/script/descriptor.cpp [hunk 6]
```diff
@@ -0,0 +1,619 @@
...
+    virtual ~PubkeyProvider(){};
```
Comment (promag) [inline src/script/descriptor.cpp]: nit, `= default`?

**Review by promag — COMMENTED** [review]
Few nits, it this for 0.17? Is `scantxoutset` considered API experimental?

File: src/script/descriptor.cpp [hunk 7]
```diff
@@ -0,0 +1,619 @@
...
+    return ret;
+}
+
+/** Split a string on every instance of sep, returning a vector. */
+std::vector<Span<const char>> Split(const Span<const char>& sp, char sep)
+{
+    std::vector<Span<const char>> ret;
+    auto it = sp.begin();
+    auto start = it;
+    while (it != sp.end()) {
```
Comment (promag) [inline src/script/descriptor.cpp]: Add
```cpp
if (it == sp.end()) return ret;
```
for when string is empty.

File: src/script/descriptor.cpp [hunk 8]
```diff
@@ -0,0 +1,619 @@
...
+    return nullptr;
+}
+
+} // namespace
+
+std::unique_ptr<Descriptor> Parse(const std::string& descriptor, FlatSigningProvider& out)
+{
+    Span<const char> sp(descriptor.data(), descriptor.size());
+    auto ret = ParseScript(sp, ParseScriptContext::TOP, out);
+    if (sp.size() == 0 && ret) return std::move(ret);
```
Comment (promag) [inline src/script/descriptor.cpp]: Compiler warning:
```
script/descriptor.cpp:617:39: warning: moving a local object in a return statement prevents copy elision [-Wpessimizing-move]
    if (sp.size() == 0 && ret) return std::move(ret);
                                      ^
script/descriptor.cpp:617:39: note: remove std::move call here
    if (sp.size() == 0 && ret) return std::move(ret);
                                      ^~~~~~~~~~  
```

File: src/rpc/blockchain.cpp [hunk 9]
```diff
@@ -1984,67 +1985,36 @@ class CoinsViewScanReserver
...
-            "        { \"address\" : \"<address>\" },       (string, optional) Bitcoin address\n"
-            "        { \"script\"  : \"<scriptPubKey>\" },  (string, optional) HEX encoded script (scriptPubKey)\n"
-            "        { \"pubkey\"  :                      (object, optional) Public key\n"
-            "          {\n"
-            "            \"pubkey\" : \"<pubkey\">,         (string, required) HEX encoded public key\n"
-            "            \"script_types\" : [ ... ],      (array, optional) Array of script-types to derive from the pubkey (possible values: \"P2PK\", \"P2PKH\", \"P2SH-P2WPKH\", \"P2WPKH\")\n"
-            "          }\n"
+            "2. \"scanobjects\"                  (array, required) Array of scan objects\n"
+            "    [                             Every scan object is either a string descriptor or an object:\n"
+            "        \"descriptor\",             (string, optional) An output descriptor\n"
```
Comment (promag) [inline src/rpc/blockchain.cpp]: Why give both options? `[{ "desc": "...", ...}, ...]` is enough?

File: src/rpc/blockchain.cpp [hunk 10]
```diff
@@ -2090,79 +2060,35 @@ UniValue scantxoutset(const JSONRPCRequest& request)
...
-                CPubKey pubkey(ParseHexV(pubkeydata_uni, "pubkey"));
-                if (!pubkey.IsFullyValid()) {
-                    throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid public key");
+            std::string desc_str;
+            int range = 1000;
+            if (scanobject.isStr()) {
+                desc_str = scanobject.get_str();
+            } else if (scanobject.isObject()) {
+                UniValue desc_uni = find_value(scanobject, "desc");
+                if (desc_uni.isNull()) throw JSONRPCError(RPC_INVALID_PARAMETER, "Descriptor needs to be provided in scan object");
```
Comment (promag) [inline src/rpc/blockchain.cpp]: Could have a test.

File: src/rpc/blockchain.cpp [hunk 11]
```diff
@@ -2090,79 +2060,35 @@ UniValue scantxoutset(const JSONRPCRequest& request)
...
+                desc_str = desc_uni.get_str();
+                UniValue range_uni = find_value(scanobject, "range");
+                if (!range_uni.isNull()) {
+                    range = range_uni.get_int();
+                    if (range < 1 || range > 1000000) throw JSONRPCError(RPC_INVALID_PARAMETER, "range out of range");
```
Comment (promag) [inline src/rpc/blockchain.cpp]: Could have a test.

File: src/script/descriptor.cpp [hunk 12]
```diff
@@ -0,0 +1,619 @@
...
+}
+
+/** Parse a script in a particular context. */
+std::unique_ptr<Descriptor> ParseScript(Span<const char>& sp, ParseScriptContext ctx, FlatSigningProvider& out)
+{
+    auto expr = Expr(sp);
+    if (Func("pk", expr)) {
+        auto pubkey = ParsePubkey(expr, ctx != ParseScriptContext::P2WSH, out);
+        if (pubkey) return MakeUnique<SingleKeyDescriptor>(std::move(pubkey), P2PKGetScript, "pk");
+    }
```
Comment (promag) [inline src/script/descriptor.cpp]: Why not return `nullptr` in these cases?

File: src/script/descriptor.h [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+     * output_script: the expanded scriptPubKeys will be put here.
+     * out: scripts and public keys necessary for solving the expanded scriptPubKeys will be put here (may be equal to provider).
+     */
+    virtual bool Expand(int pos, const SigningProvider& provider, std::vector<CScript>& output_scripts, FlatSigningProvider& out) const = 0;
+};
+
+/** Parse a descriptor string. Included private keys are put in out. Returns nullptr if parsing fails. */
+std::unique_ptr<Descriptor> Parse(const std::string& descriptor, FlatSigningProvider& out);
+
+#endif
```
Comment (promag) [inline src/script/descriptor.h]:
```
#endif // BITCOIN_SCRIPT_DESCRIPTOR_H
```

**Comment by sipa** [comment]
Rebased, removed future "key origin" support (not needed right now, ignore).

Also, I'd like people to first review the descriptor language itself (a very short summary is in the RPC help, more of a reference in script/descriptor.h, and a design document in https://gist.github.com/sipa/e3d23d498c430bb601c5bca83523fa82). In particular, I'm not convinced about the "old" function (which maps pubkeys to P2PK,P2PKH,P2WPKH, and P2SH-PWPKH) - it's very useful for dealing with existing Core wallets, but "old" doesn't really convey the right meaning. However, I don't see what else to use. "legacy" is confusing (as it would have a different meaning than legacy in -addresstype). "default" sounds like it could change over time (the intention is that it wouldn't). "sk" or "singlekey" sounds like it's encouraged and/or could change over time, ...

**Review by achow101 — COMMENTED** [review]
Concept ACK

File: src/script/descriptor.cpp (hunk 7 above)
Comment (sipa) [reply src/script/descriptor.cpp]: No, it should be `[""]`. Callers can rely on there always being at least one argument in the result.

Comment (promag) [reply src/script/descriptor.cpp]: In that case 👍

File: src/rpc/blockchain.cpp (hunk 9 above)
Comment (promag) [reply src/rpc/blockchain.cpp]: > It's pretty verbose

On the client side I think it doesn't matter (unless it's a human making the call).

**Comment by promag** [comment]
Regarding the "old" function, I think it could receive the version, for instance `default(args..., version)` — it should validate if args are compatible with the version specified (not useful at the moment).

**Comment by sipa** [comment]
@promag There is no intention of ever adding another of those "defaults". Future versions should work with specific derivations (only P2WSH directly for example, rather than a collection).

File: src/rpc/blockchain.cpp (hunk 9 above)
Comment (sipa) [reply src/rpc/blockchain.cpp]: Yes, the reason to support the short notation is to simplify things for humans.

File: src/script/descriptor.cpp [hunk 14]
```diff
@@ -0,0 +1,619 @@
...
+            if (m_derive == DeriveType::HARDENED) ret += '\'';
+        }
+        return ret;
+    }
+    bool ToPrivateString(const SigningProvider& arg, std::string& out) const override
+    {
+        CExtKey key;
+        if (!GetExtKey(arg, key)) return false;
+        out = EncodeExtKey(key);
+        for (auto entry : m_path) {
```
[the remark this answers is not in the document]
Comment (sipa) [reply src/script/descriptor.cpp]: Fixed.

File: src/rpc/blockchain.cpp [hunk 15]
```diff
@@ -1984,67 +1985,36 @@ class CoinsViewScanReserver
...
+            "2. \"scanobjects\"                  (array, required) Array of scan objects\n"
+            "    [                             Every scan object is either a string descriptor or an object:\n"
+            "        \"descriptor\",             (string, optional) An output descriptor\n"
+            "        {                         (object, optional) An object with output descriptor and metadata\n"
+            "          \"desc\": \"descriptor\",   (string, required) An output descriptor\n"
+            "          \"range\": n,             (numeric, optional) Up to what child index HD chains should be explored (default: 1000)\n"
             "        },\n"
-            "      ]\n"
+            "        ...\n"
+            "    ]\n"
```
Comment (Sjors) [inline src/rpc/blockchain.cpp]: Don't forget to add a command-line example here, it look me a while to figure out the syntax from reading the above.

I suggest an example with a typical BIP44 (legacy) xpub + BIP49 xpub would be helpful:
```
["pkh(legacy_xpub/0/*)", "pkh(legacy_xpub/1/*)", "sh(wpkh(segwit_xpub/0/*)", "sh(wpkh(segwit_xpub/1/*)"]
```

A second example could show the root master xpriv and then do hardened derivation of xpriv/44'/* or xpriv/49'/*

File: src/rpc/blockchain.cpp [hunk 16]
```diff
@@ -1984,67 +1985,36 @@ class CoinsViewScanReserver
...
         throw std::runtime_error(
             "scantxoutset <action> ( <scanobjects> )\n"
-            "\nScans the unspent transaction output set for possible entries that matches common scripts of given public keys.\n"
-            "Using addresses as scanobjects will _not_ detect unspent P2PK txouts\n"
+            "\nScans the unspent transaction output set for entries that match certain output descriptors.\n"
+            "Examples of output descriptors are:\n"
+            "    addr(<address>)                      Outputs whose scriptPubKey corresponds to the specified address (does not include P2PK)\n"
+            "    raw(<hex script>)                    Outputs whose scriptPubKey equals the specified hex scripts\n"
+            "    old(<pubkey>)                        P2PK, P2PKH, P2WPKH, and P2SH-P2WPKH outputs for the given pubkey\n"
+            "    pkh(<pubkey>)                        P2PKH outputs for the given pubkey\n"
```
Comment (Sjors) [inline src/rpc/blockchain.cpp]: `sh(wpkh())` is probably quite common, so maybe add that here?

**Review by Sjors — APPROVED** [review]
Concept ACK, and lightly tested on Ledger (P2SH-P2WPKH) and Blockchain (P2PKH) xpubs.

File: src/rpc/blockchain.cpp (hunk 15 above)
Comment (sipa) [reply src/rpc/blockchain.cpp]: Done.

File: src/rpc/blockchain.cpp (hunk 16 above)
Comment (sipa) [reply src/rpc/blockchain.cpp]: Done.

**Comment by flack** [comment]
@sipa since you write that `old` is a "combination of P2PK, P2PKH, P2WPKH, and P2SH-P2WPKH for public key P.", you could call it `combo` instead. Or `ck` if you want it to look more like the other things that use `P`. And potentially prepend `bc` to indicate it's for backwards compatibility (but in that case someone should check urban dictionary to make sure `bcck` isn't some newfangled swear word :-))

**Comment by sipa** [comment]
@flack Oh, I like that. Changed to "combo".

**Comment by MarcoFalke** [comment]
Travis failure:

```
test_framework.authproxy.JSONRPCException: Invalid descriptor 'old(03abcaca414970a8d36624011628045855cb01cded780445a33688a0b905a47224)' (-5)

**Comment by jonasschnelli** [comment]
First two findings:

* I think supporting `h` as equivalent to `'` in extended key derivation range would be great (JSON / shell / GUI). I suggest pulling in https://github.com/bitcoin/bitcoin/commit/203b4c94ce83981b0d76dd2adb778719ba868db1.

* I think the `IsHardened()` function is not correct. Using `0'/0'/*` will derive all hardened keys at the * level (same as `0'/0'/*'`). Seems to be currently impossible to derive unhardened key with the asterisk.

File: src/rpc/blockchain.cpp [hunk 17]
```diff
@@ -2090,79 +2060,35 @@ UniValue scantxoutset(const JSONRPCRequest& request)
...
-                    }
-                    assert(!script.empty());
-                    needles.insert(script);
+            FlatSigningProvider provider;
+            auto desc = Parse(desc_str, provider);
+            if (!desc) {
+                throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, strprintf("Invalid descriptor '%s'", desc_str));
+            }
+            if (!desc->IsRange()) range = 1;
+            for (int i = 0; i < range; ++i) {
```
Comment (jonasschnelli) [inline src/rpc/blockchain.cpp]: Seems like that results in an off-by-one issue (misses last key):
Suggest pulling fix: https://github.com/bitcoin/bitcoin/commit/875431149745cc517350c0620baa887c748d0b35

**Review by jonasschnelli — CHANGES_REQUESTED** [review]

**Comment by jonasschnelli** [comment]
Additional xpub range tests can be pulled from here: https://github.com/bitcoin/bitcoin/commit/cc1ffb324b53855f7eae81a6b0df10bd677ca1a5

**Comment by jonasschnelli** [comment]
@Sjors: A test for `'/*` would currently fail,... but it's simple to add (keys are already prepared in 
https://github.com/bitcoin/bitcoin/commit/cc1ffb324b53855f7eae81a6b0df10bd677ca1a5), just waiting for a fix for the hardened/non-hardened mix.

**Comment by jonasschnelli** [comment]
Suggest pulling in the experimental warning for scantxoutset here: https://github.com/bitcoin/bitcoin/commit/3ec0958ea746d0db5edefff9f423cd2caab32857

File: src/rpc/blockchain.cpp (hunk 17 above)
Comment (sipa) [reply src/rpc/blockchain.cpp]: BIP32 indexes are 0-based, so I think the current code is correct. If you expand 1000 keys, those will be indices 0 through 999.

**Comment by sipa** [comment]
@jonasschnelli Good catch; the derivation for "/*" with a parent hardened step is incorrect. Fixing.

**Comment by sipa** [comment]
@jonasschnelli Ok, fixed the derivation, changed the range to be max index rather than count, included your tests and experimental warning, and added a number of tests myself (generated with an independent BIP32 implementation).

**Comment by sipa** [comment]
@jonasschnelli Added support for h instead of ' in hardened descriptor paths as well (plus some tests for it). Serializing a descriptor will always use '.

**Review by jonasschnelli — APPROVED** [review]
Tested ACK f6b7fc349ccf9cfbeb7e91e19c20e2a2fcc9026f

**Comment by MarcoFalke** [comment]
Concept ACK f6b7fc3

**Comment by laanwj** [comment]
utACK f6b7fc349ccf9cfbeb7e91e19c20e2a2fcc9026f

File: test/functional/rpc_scantxoutset.py [hunk 18]
```diff
@@ -36,13 +52,39 @@ def run_test(self):
...
+        assert_equal(self.nodes[0].scantxoutset("start", [ "combo(tprv8ZgxMBicQKsPd7Uf69XL1XwhmjHopUGep8GuEiJDZmbQz6o58LninorQAfcKZWARbtRtfnLcJ5MQ2AtHcQJCCRUcMRvmDUjyEmNUWwx8UbK/1/1/1500)"])['total_amount'], Decimal("16.384"))
+        assert_equal(self.nodes[0].scantxoutset("start", [ "combo(tpubD6NzVbkrYhZ4WaWSyoBvQwbpLkojyoTZPRsgXELWz3Popb3qkjcJyJUGLnL4qHHoQvao8ESaAstxYSnhyswJ76uZPStJRJCTKvosUCJZL5B/1/1/0)"])['total_amount'], Decimal("4.096"))
+        assert_equal(self.nodes[0].scantxoutset("start", [ "combo(tpubD6NzVbkrYhZ4WaWSyoBvQwbpLkojyoTZPRsgXELWz3Popb3qkjcJyJUGLnL4qHHoQvao8ESaAstxYSnhyswJ76uZPStJRJCTKvosUCJZL5B/1/1/1)"])['total_amount'], Decimal("8.192"))
+        assert_equal(self.nodes[0].scantxoutset("start", [ "combo(tpubD6NzVbkrYhZ4WaWSyoBvQwbpLkojyoTZPRsgXELWz3Popb3qkjcJyJUGLnL4qHHoQvao8ESaAstxYSnhyswJ76uZPStJRJCTKvosUCJZL5B/1/1/1500)"])['total_amount'], Decimal("16.384"))
+        assert_equal(self.nodes[0].scantxoutset("start", [ {"desc": "combo(tprv8ZgxMBicQKsPd7Uf69XL1XwhmjHopUGep8GuEiJDZmbQz6o58LninorQAfcKZWARbtRtfnLcJ5MQ2AtHcQJCCRUcMRvmDUjyEmNUWwx8UbK/1/1/*')", "range": 1499}])['total_amount'], Decimal("1.536"))
```
Comment (Sjors) [inline test/functional/rpc_scantxoutset.py]: Maybe use `range: 1` and `range: 2` everywhere to speed up these tests?

**Pull request merged**